### ISSUE TITLE
[TW-796] Make task numbering persistent

### ISSUE THREAD

**Issue by taskwarrior** [GothenburgBitFactory/taskwarrior#831]
_Erlan Sergaziev on 2010-12-20T03:49:15Z says:_ 

 As far as I understand, task warrior reuses ID numbers of completed/deleted/logged task.
This creates following issues:
1.  not possible to delete/edit old erroneous or accidental tasks as they don't have ID numbers
1.  difficult to find old entries
1.  can create confusion as the user might note down (memorize) IDs of important tasks

Could you consider making task numbering continuous.
I understand that it may create long numbers, but to alleviate I can offer three possible solutions:
* automatically restart numbering at, e.g. 99 (user preference)
* provide a (at)task reset counter@ command
* use hexadecimal numbers

**Comment by taskwarrior** [comment]
Migrated metadata:
```
Created: 2010-12-20T03:49:15Z
Modified: 2014-02-09T02:06:38Z
```

**Comment by taskwarrior** [comment]
_Erlan Sergaziev on 2010-12-21T04:25:52Z says:_ 

 I have another idea on how this can be implemented.

Create one more hexadecimal field called UID whose numbering is persistent and long.
Make this field hidden by default.
Allow referencing tasks by UIDs using a special syntax, e.g.
task #UID <command>

This last syntax should work across all data files (i.e. pending and completed).

**Comment by taskwarrior** [comment]
_Federico Hernandez on 2010-12-21T04:33:18Z says:_ 

 <pre>
task #UID <command>
</pre>

not really possible as # is a special character for the shell.

**Comment by taskwarrior** [comment]
_Erlan Sergaziev on 2010-12-21T06:45:16Z says:_ 

 Yeah, you're right. Maybe then:

(at)task 0xUID < command >@

or

(at)task !UID < command >@

Anyhow, with this additional syntax it is possible to maintain both long persistent hex UID for accessing historical items and short and easy decimal ID for casual use.

One remark though, I'd still suggest that the ID counter for new tasks grows until a given threshold (e.g. 99) before starting to reuse old IDs to help alleviate 3rd problem.

**Comment by taskwarrior** [comment]
_David Patrick on 2010-12-21T11:15:44Z says:_ 

 Erlan Sergaziev wrote:

> counter for new tasks grows until a given threshold (e.g. 99) before starting to reuse old IDs to help alleviate 3rd problem.

I've been pushing for that for a while, see (rejected as dupe) #152

I have a curiosity; is it possible to take an old, done task, whose original ID number has been re-used, to re-derive the original ID number from the original UUID , anyways ? (not to re-assign it, or anything, but for historical reference) If this is possible, then it would obviate the need for a new (and potentially confusing) reference number.

UUID is the canonical reference for any task, done or not, and if users could search for uuid (done or not), something like

 % task any uuid:f7f711a5-abb0-6b6b-27b8-c584f7e51bb0

or even a piece of that, perhaps the most significant part (?)

 % task any uuid.has:e51bb0

then at least long term ID reference could be maintained

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2010-12-30T20:37:36Z says:_ 

 As far as making IDs permanent, I'm against it.

I would be willing to make tasks addressable by UUID, or a unique leading fragment of a UUID.  This has long been a goal, although a low priority one.  This would therefore be a taskwarrior 2.x feature, because we really need that new parser integrated before we start messing with the command line.

Here is the syntax I would implement:

<pre>
task <id> <command>
</pre>

I would allow the ID to be either a sequence ID as it is now, or a certain minimum number of leading characters from a UUID.  No new syntax necessary.

**Comment by taskwarrior** [comment]
_Erlan Sergaziev on 2010-12-30T22:26:18Z says:_ 

 Paul,

What you are suggesting will help many, I am sure.

Myself I have since found a workaround - I found out that tasks are stored in a plain text format.
Because I am pretty good at VIM, it wasn't difficult at all to operate on old tasks.

Thanks and Happy New Year!

Erlan

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2010-12-30T22:55:17Z says:_ 

 Erlan, I'm glad you have a workaround - it eases the pressure.  Be careful when editing that file!

Best wishes and happy new year to you too.
P.

**Comment by taskwarrior** [comment]
_steve rader on 2011-01-02T09:37:16Z says:_ 

 FWIW: I ALMOST passed over task because of the task numbering.  It really seemed like
a bug/problem. But then I realized: it's a great feature because seeing very
small task numbers provides subtle encouragement that you're getting things done!
Perhaps that's a unspoken philosophy which ought to be easily encountered by folks
just thinking/glancing at task.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2011-01-02T11:17:40Z says:_ 

 Steve, it was never intentional that low ID numbers provide encouragement, but if it does, then good!

I think we need to put together an eloquent justification for the current system, because this question comes up a lot.  Such a document will include:

* Shorter id numbers means fewer data entry errors (I have mistype 236, 263, 326 many times, and it's hard to spot).
* Complexity overwhelms.  One of the reasons we categorize is so that we can focus on small lists at a time.  I have 300+ pending, and 2000+ completed tasks.  If ID numbers were permanent, all my tasks would have 4-digit ids.  Awful.

That said, Erlan will get his UUID-addressable tasks in 2.0.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2011-04-18T05:36:54Z says:_ 

 Note: setting rc.gc=off essentially makes ID numbers persistent.  Note that this means performance will degrade linearly with the total nubmer of tasks (pending + completed + delete).

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2011-05-14T04:19:13Z says:_ 

 We have a workaround, I don't want to make a permanent change, so I'm closing.  Let me know if there are strong objections to this.
